test/task.h

```diff
@@ -109,6 +109,134 @@ typedef enum {
+#define ASSERT_INT_GT(a, b) ASSERT_INT_BASE(a, >, b, int, i)
+#define ASSERT_INT_LE(a, b) ASSERT_INT_BASE(a, <=, b, int, i)
+#define ASSERT_INT_LT(a, b) ASSERT_INT_BASE(a, <, b, int, i)
+#define ASSERT_INT_NE(a, b) ASSERT_INT_BASE(a, !=, b, int, i)
+
+#define ASSERT_UINT_EQ(a, b) ASSERT_INT_BASE(a, ==, b, unsigned int, u)
+#define ASSERT_UINT_GE(a, b) ASSERT_INT_BASE(a, >=, b, unsigned int, u)
+#define ASSERT_UINT_GT(a, b) ASSERT_INT_BASE(a, >, b, unsigned int, u)
+#define ASSERT_UINT_LE(a, b) ASSERT_INT_BASE(a, <=, b, unsigned int, u)
+#define ASSERT_UINT_LT(a, b) ASSERT_INT_BASE(a, <, b, unsigned int, u)
+#define ASSERT_UINT_NE(a, b) ASSERT_INT_BASE(a, !=, b, unsigned int, u)
+
+#define ASSERT_LONG_INT_EQ(a, b) ASSERT_INT_BASE(a, ==, b, long int, li)
+#define ASSERT_LONG_INT_GE(a, b) ASSERT_INT_BASE(a, >=, b, long int, li)
+#define ASSERT_LONG_INT_GT(a, b) ASSERT_INT_BASE(a, >, b, long int, li)
+#define ASSERT_LONG_INT_LE(a, b) ASSERT_INT_BASE(a, <=, b, long int, li)
+#define ASSERT_LONG_INT_LT(a, b) ASSERT_INT_BASE(a, <, b, long int, li)
+#define ASSERT_LONG_INT_NE(a, b) ASSERT_INT_BASE(a, !=, b, long int, li)
+
+#define ASSERT_LONG_UINT_EQ(a, b) ASSERT_INT_BASE(a, ==, b, unsigned long int, lu)
+#define ASSERT_LONG_UINT_GE(a, b) ASSERT_INT_BASE(a, >=, b, unsigned long int, lu)
+#define ASSERT_LONG_UINT_GT(a, b) ASSERT_INT_BASE(a, >, b, unsigned long int, lu)
+#define ASSERT_LONG_UINT_LE(a, b) ASSERT_INT_BASE(a, <=, b, unsigned long int, lu)
+#define ASSERT_LONG_UINT_LT(a, b) ASSERT_INT_BASE(a, <, b, unsigned long int, lu)
+#define ASSERT_LONG_UINT_NE(a, b) ASSERT_INT_BASE(a, !=, b, unsigned long int, lu)
+
+#define ASSERT_LONG_LONG_INT_EQ(a, b) ASSERT_INT_BASE(a, ==, b, long long int, lli)
+#define ASSERT_LONG_LONG_INT_GE(a, b) ASSERT_INT_BASE(a, >=, b, long long int, lli)
+#define ASSERT_LONG_LONG_INT_GT(a, b) ASSERT_INT_BASE(a, >, b, long long int, lli)
+#define ASSERT_LONG_LONG_INT_LE(a, b) ASSERT_INT_BASE(a, <=, b, long long int, lli)
+#define ASSERT_LONG_LONG_INT_LT(a, b) ASSERT_INT_BASE(a, <, b, long long int, lli)
+#define ASSERT_LONG_LONG_INT_NE(a, b) ASSERT_INT_BASE(a, !=, b, long long int, lli)
+
+#define ASSERT_LONG_LONG_UINT_EQ(a, b) ASSERT_INT_BASE(a, ==, b, unsigned long long int, llu)
+#define ASSERT_LONG_LONG_UINT_GE(a, b) ASSERT_INT_BASE(a, >=, b, unsigned long long int, llu)
+#define ASSERT_LONG_LONG_UINT_GT(a, b) ASSERT_INT_BASE(a, >, b, unsigned long long int, llu)
+#define ASSERT_LONG_LONG_UINT_LE(a, b) ASSERT_INT_BASE(a, <=, b, unsigned long long int, llu)
+#define ASSERT_LONG_LONG_UINT_LT(a, b) ASSERT_INT_BASE(a, <, b, unsigned long long int, llu)
+#define ASSERT_LONG_LONG_UINT_NE(a, b) ASSERT_INT_BASE(a, !=, b, unsigned long long int, llu)
+
+#define ASSERT_STR_EQ(a, b) \
+  ASSERT_BASE(strcmp(a, b) == 0, a, ==, b, s)
+
+#define ASSERT_STR_NE(a, b) \
+  ASSERT_BASE(strcmp(a, b) != 0, a, !=, b, s)
+
+#define ASSERT_MEM_EQ(a, b, size) \
+  ASSERT_BASE_LEN(memcmp(a, b, size) == 0, a, ==, b, s, size)
```

Comment:
Wouldn't you always want to dump the contents of `a` and `b` on inequality?

Reply:
Let me rephrase: `memcmp()` implies we're comparing binary data, not strings, so does it make sense to try to print them as strings?

(I realize that most of the time we're comparing buffers containing ASCII text so the answer may well be "yes, it does.")

Perhaps a middle ground is an `xxd()` like function that dumps hex and text? I'm sure I have one in one of my projects that I could dig up. Doesn't have to hold up this PR though.

Reply:
That's the reason I created the `HEX` versions for the case of binary data. Anyway, if you could share some code, it would be great.